### PR TITLE
Workaround flow parser bug with spread in arrays

### PR DESCRIPTION
**Pull request by vjeux**
The current output of

```js
[...a, ...b]
```

is

```js
[...a, , ...b]
```

because flow parses it as

```
ArrayExpression(SpreadExpression, null, SpreadExpression)
```

This is a bug in the flow parser. Until it gets fixed, we can workaround it by deleting the `null` after a `SpreadExpression`.

### CODE REVIEW

**Comment by vjeux** [comment]
FYI, this has already been fixed in master and will ship with flow 0.38. It was also "broken" due to a bad revert ( https://github.com/facebook/flow/commit/f6b628098bcb750fd206ddc739b39656ec69ed1b ) and isn't an issue with flow 0.36. I think we should land this and whenever the next version of flow ships, we can just get rid of this block.

This workaround shouldn't cause too much trouble even with the fix in. It's going to cause issues if you write `[...a, , b]` but in practice it is extremely rare so we can live with it if we forget.

**Comment by jlongster** [comment]
The code `[...a, ...b]` is invalid isn't it? As far as I know you can only have one spread element and it has to be the last one.

I did noticed these kinds of issues on nuclide though -- it was adding extra commas. I can't remember if they all involved spread elements.

File: src/printer.js [hunk 1]
```diff
@@ -641,6 +641,24 @@ function genericPrintNoParens(path, options, print) {
...
         parts.push("[]");
       } else {
+        // The flow parser has a bug where
+        //   [ ...a, ...a ]
+        // is being parsed as
+        //   ArrayExpression(SpreadElement, null, SpreadElement)
+        // The issue is that it treats the comma as a null element.
+        //
+        // In order to workaround this bug, we delete the null of a sequence
+        // [SpreadElement, null].
```
Comment (jlongster) [inline src/printer.js]: Can you add that we can remove this code when we upgrade to Flow 0.38?

**Comment by lydell** [comment]
> The code `[...a, ...b]` is invalid isn't it? As far as I know you can only have one spread element and it has to be the last one.

You can only have a single _rest_ element (in destructuring and parameter lists), but multiple spread elements is fine (in arrays and when calling functions).

**Comment by vjeux** [comment]
I added the comment to remove in 0.38.

For `[...a, ...b]`, @lydell is right, when you are trying to destructure things, you cannot have more than one rest otherwise it is ambiguous. But when you are creating things, having multiple just means you are concatenating all of them together, it is fine.

All the nuclide issues are because of this. We use those pretty heavily :)

**Comment by jlongster** [comment]
Oh, of course. Thanks a lot!

**Comment by vjeux** [comment]
Oh noes, flow 0.38 was just released a few minutes ago :p
https://twitter.com/flowtype/status/821766587661352963

**Pull request merged**